vms/platformvm/spend.go

```diff
@@ -350,204 +524,237 @@ func (vm *VM) semanticVerifySpend(
 		if err != nil {
-			return fmt.Errorf("couldn't marshal owner: %w", err)
+			return err
 		}
-		ownerID := hashing.ComputeHash256Array(ownerBytes)
-		owners, ok := lockedConsumed[locktime]
+		ownerConsumedAmounts[utxoLockIDs] = newConsumedAmounts
+
+		// cause we will iterate over producedAmounts and check against consumed
+		// and we also need to keep track of unlocked burned amont
+		// we want to have all consumed ownerIDs/LockIDs in place
+
+		ownerProducedAmounts, ok := producedAmounts[ownerID]
 		if !ok {
-			owners = make(map[ids.ID]uint64)
-			lockedConsumed[locktime] = owners
-		}
-		newAmount, err := math.Add64(owners[ownerID], amount)
-		if err != nil {
-			return err
+			ownerProducedAmounts = map[LockIDs]uint64{}
+			producedAmounts[ownerID] = ownerProducedAmounts
 		}
-		owners[ownerID] = newAmount
+		ownerProducedAmounts[utxoLockIDs] = 0
 	}
 
-	for _, out := range outs {
-		if assetID := out.AssetID(); assetID != feeAssetID {
+	for _, output := range outputs {
+		if output.AssetID() != assetID {
 			return errAssetIDMismatch
 		}
-
-		output := out.Output()
-		locktime := uint64(0)
-		// Set [locktime] to this output's locktime, if applicable
-		if inner, ok := output.(*StakeableLockOut); ok {
-			output = inner.TransferableOut
-			locktime = inner.Locktime
+		out := output.Out
+		lockIDs := LockIDs{}
+		if lockedOut, ok := out.(*LockedOut); ok {
+			lockIDs = lockedOut.LockIDs
+			out = lockedOut.TransferableOut
 		}
 
-		amount := output.Amount()
-
-		if locktime == 0 {
-			newUnlockedProduced, err := math.Add64(unlockedProduced, amount)
-			if err != nil {
-				return err
-			}
-			unlockedProduced = newUnlockedProduced
-			continue
-		}
-
-		owned, ok := output.(Owned)
-		if !ok {
-			return errUnknownOwners
-		}
-		owner := owned.Owners()
-		ownerBytes, err := Codec.Marshal(CodecVersion, owner)
+		ownerID, err := getOwnerID(out)
 		if err != nil {
-			return fmt.Errorf("couldn't marshal owner: %w", err)
+			return err
 		}
-		ownerID := hashing.ComputeHash256Array(ownerBytes)
-		owners, ok := lockedProduced[locktime]
+
+		ownerProducedAmounts, ok := producedAmounts[ownerID]
 		if !ok {
-			owners = make(map[ids.ID]uint64)
-			lockedProduced[locktime] = owners
+			ownerProducedAmounts = map[LockIDs]uint64{}
+			producedAmounts[ownerID] = ownerProducedAmounts
 		}
-		newAmount, err := math.Add64(owners[ownerID], amount)
+
+		newProducedAmount, err := math.Add64(ownerProducedAmounts[lockIDs], out.Amount())
 		if err != nil {
 			return err
 		}
-		owners[ownerID] = newAmount
+		ownerProducedAmounts[lockIDs] = newProducedAmount
 	}
 
-	// Make sure that for each locktime, tokens produced <= tokens consumed
-	for locktime, producedAmounts := range lockedProduced {
-		consumedAmounts := lockedConsumed[locktime]
-		for ownerID, producedAmount := range producedAmounts {
-			consumedAmount := consumedAmounts[ownerID]
-
+	unlockedBurned := uint64(0)
```

Comment:
I don't see any need to handle any(!!) ownership thing here.
This function has to be reworked completely.